language/move-compiler/src/hlir/ast.rs

```diff
@@ -436,6 +436,7 @@ impl BaseType_ {
                 )
                 .unwrap()
             }
+            Fun => panic!("ICE unexpected function type"),
```

Comment:
> I don't understand why an explicit variance enum was added. Variance already appears through the fact that `join` cares about the order of arguments when it has `TypingCase::Subtype`
> 
> You should be able to express everything you need for variance by flipping the arguments

You are really comparing in join (A -> B) <: (A' -> B'), so the variance is needed if you do not want to special case the function type constructor in join. This can indeed be done, but I find it cleaner to just have a concept of variance.

Reply:
> will this be caught by the unimplemented diag?

Should be, otherwise there is in fact an internal error.